packages/next/src/lib/metadata/generate/basic.tsx

```diff
@@ -1,15 +1,47 @@
 import type { ResolvedMetadata } from '../types/metadata-interface'
+import type { Viewport } from '../types/extra-types'
 
 import React from 'react'
 import { Meta } from './utils'
 
+const viewPortKeys = {
+  width: 'width',
+  height: 'height',
+  initialScale: 'initial-scale',
+  minimumScale: 'minimum-scale',
+  maximumScale: 'maximum-scale',
+  viewportFit: 'viewport-fit',
+} as const
+
+function ViewPortMeta({
+  viewport,
+}: {
+  viewport: ResolvedMetadata['viewport']
+}) {
+  let content
+  if (typeof viewport === 'string') {
+    content = viewport
+  } else if (viewport) {
+    content = ''
+    for (const key_ in viewPortKeys) {
+      const key = key_ as keyof Viewport
+      if (viewport[key]) {
+        if (content) content += ', '
+        content += `${viewPortKeys[key]}=${viewport[key]}`
+      }
+    }
+  }
+  return <Meta name="viewport" content={content} />
+}
```

Comment:
We need to move this to the resolution stage. We accept a viewport string or object but since every type is going to have just 1 resolved type, in this case we're going to use 'string'. This is because parsing it to object may be lossy (maybe folks want to use the some legacy options for instance). So if we receive a Viewport object in Metadata we need to construct it as a string and save that to ResolvedMetadata. When we render it can just be a regular meta tag